lib/ruby-push-notifications/mpns/mpns_connection.rb

```diff
@@ -0,0 +1,72 @@
+require 'uri'
+require 'net/https'
+
+module RubyPushNotifications
+  module MPNS
+    # Encapsulates a connection to the MPNS service
+    # Responsible for final connection with the service.
+    #
+    class MPNSConnection
+
+      # @private Content-Type HTTP Header string
+      CONTENT_TYPE_HEADER  = 'Content-Type'
+
+      # @private text/xml content type
+      X_NOTIFICATION_CLASS = 'X-NotificationClass'
+
+      # @private Windows Phone Target
+      X_WINDOWSPHONE_TARGET = 'X-WindowsPhone-Target'
+
+      # @private text/xml content type
+      XML_CONTENT_TYPE     = 'text/xml'
+
+      # @private Enumators for notification types
+      BASEBATCH = { tile: 1, toast: 2, raw: 3 }
+
+      # @private Enumators for delay
+      BATCHADDS = { delay450: 10, delay900: 20 }
+
+      # @private Windows Phone Target Types
+      WP_TARGETS = { toast: 'toast', tile: 'token' }
+
+      # Issues a POST request to the MPNS send endpoint to
+      # submit the given notifications.
+      #
+      # @param n [MPNSNotification]. The notification object to POST
+      # @param cert [String]. Contents of the PEM encoded certificate
+      # @return [Array]. The response of post
+      # (http://msdn.microsoft.com/pt-br/library/windows/apps/ff941099)
+      def self.post(n, cert)
+        notification_class = calculate_delay n.data[:type], n.data[:delay]
+        headers = {
+          CONTENT_TYPE_HEADER => XML_CONTENT_TYPE,
+          X_NOTIFICATION_CLASS => notification_class.to_s
+        }
+        headers[X_WINDOWSPHONE_TARGET] = WP_TARGETS[n.data[:type]] unless n.data[:type] == :raw
+        body = n.as_mpns_xml
+        responses = []
+        n.device_urls.each do |url|
```

Comment:
I don't think the connection needs to know how to extract the device_urls from notifications. I'd rather see something like: `n.each_device...`